src/adapter/src/catalog/apply.rs

```diff
@@ -799,17 +807,126 @@ impl CatalogState {
+        let conn_id = old_entry.item().conn_id().unwrap_or(&SYSTEM_CONN_ID);
+        let schema = self.get_schema_mut(
+            &old_entry.name().qualifiers.database_spec,
+            &old_entry.name().qualifiers.schema_spec,
+            conn_id,
+        );
+        schema.items.remove(&old_entry.name().item);
+
+        // Dropped deps
+        let dropped_references: Vec<_> = old_entry
+            .references()
+            .0
+            .difference(&item.references().0)
+            .cloned()
+            .collect();
+        let dropped_uses: Vec<_> = old_entry.uses().difference(&item.uses()).cloned().collect();
+
+        // We only need to install this item on items in the `referenced_by` of new
+        // dependencies.
+        let new_references: Vec<_> = item
+            .references()
+            .0
+            .difference(&old_entry.references().0)
+            .cloned()
+            .collect();
+        // We only need to install this item on items in the `used_by` of new
+        // dependencies.
+        let new_uses: Vec<_> = item.uses().difference(&old_entry.uses()).cloned().collect();
+
+        let mut new_entry = old_entry;
+        new_entry.item = item;
+        new_entry.id = id;
+        new_entry.oid = oid;
+        new_entry.name = name;
+        new_entry.owner_id = owner_id;
+        new_entry.privileges = privileges;
+
+        schema.items.insert(new_entry.name().item.clone(), id);
+
+        for u in dropped_references {
+            // OK if we no longer have this entry because we are dropping our
+            // dependency on it.
+            if let Some(metadata) = self.entry_by_id.get_mut(&u) {
+                metadata.referenced_by.retain(|dep_id| *dep_id != id)
+            }
+        }
+
+        for u in dropped_uses {
+            // OK if we no longer have this entry because we are dropping our
+            // dependency on it.
+            if let Some(metadata) = self.entry_by_id.get_mut(&u) {
+                metadata.used_by.retain(|dep_id| *dep_id != id)
+            }
+        }
+
+        for u in new_references {
+            match self.entry_by_id.get_mut(&u) {
+                Some(metadata) => metadata.referenced_by.push(new_entry.id()),
+                None => panic!(
+                    "Catalog: missing dependent catalog item {} while updating {}",
+                    &u,
+                    self.resolve_full_name(new_entry.name(), new_entry.conn_id())
+                ),
+            }
+        }
+        for u in new_uses {
+            match self.entry_by_id.get_mut(&u) {
+                Some(metadata) => metadata.used_by.push(new_entry.id()),
+                None => panic!(
+                    "Catalog: missing dependent catalog item {} while updating {}",
+                    &u,
+                    self.resolve_full_name(new_entry.name(), new_entry.conn_id())
+                ),
+            }
+        }
+
+        self.entry_by_id.insert(id, new_entry);
+    }
```

Comment:
This is 99% copied from `update_item` in `transact.rs` The one difference is that I added the following lines to make it more general purpose:
```
        new_entry.id = id;
        new_entry.oid = oid;
        new_entry.owner_id = owner_id;
        new_entry.privileges = privileges;
```